txnkv/transaction/pessimistic.go

```diff
@@ -150,156 +168,375 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
 		sender := locate.NewRegionRequestSender(c.store.GetRegionCache(), c.store.GetTiKVClient())
 		startTime := time.Now()
 		resp, err := sender.SendReq(bo, req, batch.region, client.ReadTimeoutShort)
-		reqDuration := time.Since(startTime)
+		diagCtx.reqDuration = time.Since(startTime)
+		diagCtx.sender = sender
 		if action.LockCtx.Stats != nil {
-			atomic.AddInt64(&action.LockCtx.Stats.LockRPCTime, int64(reqDuration))
+			atomic.AddInt64(&action.LockCtx.Stats.LockRPCTime, int64(diagCtx.reqDuration))
 			atomic.AddInt64(&action.LockCtx.Stats.LockRPCCount, 1)
 		}
 		if err != nil {
 			return err
 		}
-		regionErr, err := resp.GetRegionError()
-		if err != nil {
-			return err
-		}
-		if regionErr != nil {
-			// For other region error and the fake region error, backoff because
-			// there's something wrong.
-			// For the real EpochNotMatch error, don't backoff.
-			if regionErr.GetEpochNotMatch() == nil || locate.IsFakeRegionError(regionErr) {
-				err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
-				if err != nil {
-					return err
-				}
+
+		if action.wakeUpMode == kvrpcpb.PessimisticLockWakeUpMode_WakeUpModeNormal {
+			finished, err := action.handlePessimisticLockResponseNormalMode(c, bo, &batch, mutations, resp, &diagCtx)
+			if err != nil {
+				return err
+			}
+			if finished {
+				return nil
 			}
-			same, err := batch.relocate(bo, c.store.GetRegionCache())
+		} else if action.wakeUpMode == kvrpcpb.PessimisticLockWakeUpMode_WakeUpModeForceLock {
+			finished, retryMutations, err := action.handlePessimisticLockResponseForceLockMode(c, bo, &batch, mutations, resp, &diagCtx)
 			if err != nil {
 				return err
 			}
-			if same {
-				continue
+			if finished {
+				return nil
+			}
+
+			if retryMutations != nil && retryMutations.Len() < batch.mutations.Len() {
+				// The request may be partially succeeded. Replace the request with remaining mutations.
+				// The primary lock (if it's in this request) must have been succeeded.
```

Comment:
Where is it ensured? (Though it's not important when we only enable this mode for single key)

Reply:
TiKV should ensure it. But since this kind of code is useless in current version, I'm now removing them.